mqtt/mqtt-bridge/src/persist/waking_state/ring_buffer/mod.rs

```diff
@@ -353,248 +427,293 @@ impl RingBuffer {
     fn drop(&mut self) {
-        let result = self.mmap.flush();
+        let result = self.file.flush();
         if let Some(err) = result.err() {
             error!("Failed to flush {:?}", err);
         }
+        let result = self.metadata_file.flush();
+        if let Some(err) = result.err() {
+            error!("Failed to flush metadata {:?}", err);
+        }
     }
 }
 
-fn create_file(file_path: &Path) -> IOResult<File> {
+fn create_file(file_path: &PathBuf) -> IOResult<File> {
+    if let Some(parent) = file_path.parent() {
+        create_dir_all(parent)?;
+    }
     OpenOptions::new()
         .read(true)
         .write(true)
         .create(true)
         .open(file_path)
 }
 
+fn save_ring_buffer_metadata(metadata: &RingBufferMetadata, file: &mut File) -> BincodeResult<()> {
+    let buf = bincode::serialize(metadata)?;
+    file.seek(SeekFrom::Start(0))?;
+    file.write_all(&buf)?;
+    file.flush()?;
+    Ok(())
+}
+
+fn retrieve_ring_buffer_metadata(file: &mut File) -> BincodeResult<RingBufferMetadata> {
+    let mut buf = vec![];
+    file.read_to_end(&mut buf)?;
+    if buf.is_empty() {
+        return Ok(RingBufferMetadata::default());
+    }
+    bincode::deserialize::<RingBufferMetadata>(&buf)
+}
+
 fn find_pointers_and_order_post_crash(
-    mmap: &MmapMut,
-    max_file_size: u32,
-) -> StorageResult<(FilePointers, u128)> {
-    let mut block: BlockHeaderWithHash;
+    // mmap: &MmapMut,
+    file: &mut File,
+    max_file_size: u64,
+    best_guess_metadata: RingBufferMetadata,
+) -> RingBufferMetadata {
+    let block_size = *SERIALIZED_BLOCK_SIZE;
+
+    let mut start = best_guess_metadata.file_pointers.write;
+    let mut end = start + block_size;
+    let mut read = best_guess_metadata.file_pointers.read_begin;
+    let mut write = best_guess_metadata.file_pointers.write;
+    let mut order = best_guess_metadata.order;
+    let can_read_from_wrap_around = best_guess_metadata.can_read_from_wrap_around;
+
+    // Now that a block has been found, we can find the last write.
+    // We can scan each block we find and check the order on it, if
+    // we find a block that has any of the following we can stop searching:
+    // 1. An order that is less than current block (wrap-around)
+    // 2. All zero memory, which is nothing was written
+    // 3. Failure to load block, which could be data corruption, but we can
+    // at least serve new data.
+    // Note read pointer is found a different way.
+    // Read is found by scanning blocks marked for overwrite
+    // and taking the value of the first non-overwrite block.
 
+    // if we have more write_updates than read_updates and write == read
+    // then we must be in wrap_around case.
+    let mut write_updates = 0;
+    let mut read_updates = 0;
     #[allow(clippy::cast_possible_truncation)]
-    let block_size = *SERIALIZED_BLOCK_SIZE as u32;
+    let mut block = match load_block_header(file, start, block_size as usize, max_file_size) {
```

Comment:
This will not work as we are short-circuiting the method with the return and not for assigning to the block. I am okay to refactoring this into something else though :) Just let me know!